protos/tunneloffload.proto

```diff
@@ -0,0 +1,352 @@
+  For details of each fields, please refer to the RFC
+
+  Notes:
+  - Version field is not present since always 0
+  */  
+
+  // IP's that will be in the outer IP of encapsulated packet
+  oneof ip {
+    IPV4Pair outerIpv4Pair = 1;
+    IPV6Pair outerIpv6Pair = 2;
+  }
+
+  // Source & Dest mac of inner GENEVE packet 
+  MacPair innerMacPair = 3;
+
+  uint32 optionLength = 4; // 6 bits 
+  bool controlPacket = 5; // O bit at rfc. True is '1', False is '0'. Default is False.
+  bool criticalOptionPresent = 6; // C bit at rfc. True is '1', False is '0'. Default is False.
+  uint32 vni = 7; // Virtual Network Identifier - 24 bits
+  repeated GENEVEOption geneveOption = 8;
+}
+
+message GENEVEDecap {
+  // GENEVEDecap can only be used if GENEVE 
+  // is on the match of the tunnel
+}
+
+message NATAction {
+  fixed32 sourceIp = 1;
+}
+
+message MacPair {
+  bytes sourceMac = 1;
+  bytes destinationMac = 2;
+}
+
+message IPV4Pair {  
+  fixed32 sourceIp = 1;
+  fixed32 destinationIp = 2;
+}
+
+message IPV6Pair {
+  bytes sourceIp = 1;
+  bytes destinationIp = 2; 
+}
+
+message IPV4Match {
+  fixed32 sourceIp = 1;
+  uint32 sourceIpPrefix = 2;
+  fixed32 destinationIp = 3;
+  uint32 destinationIpPrefix = 4;
+}
+
+message IPV6Match {
+  bytes sourceIp = 1;
+  uint32 sourceIpPrefix = 2;
+  bytes destinationIp = 3;
+  uint32 destinationIpPrefix = 4;
+}
+
+enum IPSecTunnelType {
+  TRANSPORT = 0;
+  TUNNEL = 1;
+  TRANSPORT_NAT_TRAVERSAL = 2; 
+  TUNNEL_NAT_TRAVERSAL = 3;
+}
+
+message IPSecEnc {
+  IPSecTunnelType tunnelType = 1;
+  ENC_TYPE encryptionType = 2;
+  bytes authenticationKey = 3;
+  AUTH_TYPE authenticationType = 4;
+  bytes encryptionKey = 5;
+  // SPI should be unique for every session
+  uint32 SPI = 6;
+  // In case the ipsec tunnel is with tunnel mode,
+  // one of the messages below should present
+  oneof tunnelIps {
+    IPV4Pair ipv4_tunnel = 7;
+    IPV6Pair ipv6_tunnel = 8;
+  }
+}
+
+message IPSecDec {
+  IPSecTunnelType tunnelType = 1;
+  ENC_TYPE encryptionType = 2;
+  bytes authenticationKey = 3;
+  AUTH_TYPE authenticationType = 4;
+  bytes encryptionKey = 5;
+  oneof tunnelIps {
+    IPV4Pair ipv4_tunnel = 6;
+    IPV6Pair ipv6_tunnel = 7;
+  }
+  
+}
+
+message IPSecTunnel {
+  oneof ipsec {
+    IPSecEnc ipsecEnc = 1;
+    IPSecDec ipsecDec = 2;
+  }
+
+}
+
+message tunnelId {
+  uint64 tunnelId = 1;
+}
+
+message counters {
+  uint64 inPackets = 1;
+  uint64 outPackets = 2;
+  uint64 inBytes  = 3;
+  uint64 outBytes = 4;
+  uint64 inPacketsDrops = 5;
+  uint64 outPacketsDrops = 6;
+  uint64 inBytesDrops  = 7;
+  uint64 outBytesDrops = 8;
+
+}
+
+message ipTunnelResponses {
+  ADD_TUNNEL_STATUS requestStatus = 1;
+  uint64 errorStatus = 2;
+  repeated ipTunnelResponse responses = 3;
+}
+
+message ipTunnelResponse {
+  // Tunnel ID assigned to this tunnel
+  uint64 tunnelId = 1;
+
+  // Information regards the ipTunnel (including match, tunnel information)
+  ipTunnelRule ipTunnel = 2; 
```

Comment:
rules should be in getIpTunnels not in response to a addIpTunnelRequest - what is this a response to ?